### ISSUE TITLE
Managed to build a sub 90mb (compressed) docker image.

### ISSUE THREAD

**Issue by battletux** [juice-shop/juice-shop#404]
So using a multi stage dockerfile and dockers auto build system I have managed to get a docker image to 87mb (compressed).

I've checked with a couple of the simple challenges (XSS Tier 1, Error Handling, Login Bender) and it seems to work.

Could someone check some of the more complex challenges to confirm it all works before I submit a pull request please?

`docker run -d -p 3000:3000 arclight/juice-shop_x64_test`

Thanks.

**Comment by J12934** [comment]
That's really cool! 🎉
I ran the API Tests against the container and it looks like everything is working fine.

Where does this size improvements coming from? 
npm cache and the git package?

**Comment by battletux** [comment]
I used a multi stage build, by adding in a 2nd stage to your existing Dockerfile:

```
# OWASP Juice Shop - An intentionally insecure JavaScript Web Application
FROM            node:6-alpine
MAINTAINER      Bjoern Kimminich <bjoern.kimminich@owasp.org>
LABEL version = "5.0.3"

RUN apk update && apk add git

COPY . /juice-shop
WORKDIR /juice-shop
RUN npm install --production --unsafe-perm

# 2nd stage code here:

FROM    node:6-alpine
WORKDIR /juice-shop
COPY --from=0 /juice-shop .

# End of 2nd stage code

EXPOSE  3000
CMD ["npm", "start"]

```

My understanding is that docker builds the container as normal  untill it gets to the 2nd FROM line. It then loads a new, clean, container and creates the WORKDIR in this new container and copies the installed juice-shop from the WORKDIR of the first container into the WORKDIR of this new container.

The reason this saves space, at least for my understanding, is that there are no left over un-needed tools from the build process (which you would have in a traditional single stage build).

Not sure how the automated build system manages to get it this small as when I build it manually on my mac it is still over 150mb in size.

If you're happy that it works I'll raise the pull request.

On another note I've updated the raspberry pi docker image to be the latest release (5.0.3) and have also had some success in reducing it's file size in a similar way, 416mb (latest tag) down to 168mb (beta2 tag). 

I'm still trying to figure out a multi-arch docker build so that it can all be built in one go for both architectures.

**Comment by bkimminich** [comment]
Could you fire up a server with the latest shrunk Docker and then run the e2e tests directly*  with `protractor protractor.conf.js`? If that passes, every challenge should work fine!

_*instead of `npm run protractor` which starts a server from the work directory_

**Comment by battletux** [comment]
Sure, I'll get that running now.

_edit_
Seems protractor is not installed so having to run `npm run protractor`.

**Comment by battletux** [comment]
Ok, seems npm run protractor is failing:

```
/juice-shop # npm run protractor

> juice-shop@5.0.3 preprotractor /juice-shop
> npm run update-webdriver

> juice-shop@5.0.3 preupdate-webdriver /juice-shop
> npm install

npm WARN optional SKIPPING OPTIONAL DEPENDENCY: iltorb@^1.0.13 (node_modules/grunt-contrib-compress/node_modules/iltorb):
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: Error: connect ECONNREFUSED 151.101.16.162:443
npm ERR! Linux 4.9.49-moby
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install"
npm ERR! node v6.12.0
npm ERR! npm  v3.10.10
npm ERR! code ECONNREFUSED
npm ERR! errno ECONNREFUSED
npm ERR! syscall connect

npm ERR! Error: connect ECONNREFUSED 151.101.16.162:443
npm ERR!     at Object.exports._errnoException (util.js:1020:11)
npm ERR!     at exports._exceptionWithHostPort (util.js:1043:20)
npm ERR!     at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1090:14)
npm ERR!  { Error: connect ECONNREFUSED 151.101.16.162:443
npm ERR!     at Object.exports._errnoException (util.js:1020:11)
npm ERR!     at exports._exceptionWithHostPort (util.js:1043:20)
npm ERR!     at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1090:14)
npm ERR!   code: 'ECONNREFUSED',
npm ERR!   errno: 'ECONNREFUSED',
npm ERR!   syscall: 'connect',
npm ERR!   address: '151.101.16.162',
npm ERR!   port: 443,
npm ERR!   parent: 'juice-shop' }
npm ERR!
npm ERR! If you are behind a proxy, please make sure that the
npm ERR! 'proxy' config is set properly.  See: 'npm help config'

npm ERR! Please include the following file with any support request:
npm ERR!     /juice-shop/npm-debug.log

npm ERR! Linux 4.9.49-moby
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "run" "update-webdriver"
npm ERR! node v6.12.0
npm ERR! npm  v3.10.10
npm ERR! code ELIFECYCLE
npm ERR! juice-shop@5.0.3 preupdate-webdriver: `npm install`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the juice-shop@5.0.3 preupdate-webdriver script 'npm install'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the juice-shop package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     npm install
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs juice-shop
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls juice-shop
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /juice-shop/npm-debug.log

npm ERR! Linux 4.9.49-moby
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "run" "protractor"
npm ERR! node v6.12.0
npm ERR! npm  v3.10.10
npm ERR! code ELIFECYCLE
npm ERR! juice-shop@5.0.3 preprotractor: `npm run update-webdriver`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the juice-shop@5.0.3 preprotractor script 'npm run update-webdriver'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the juice-shop package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     npm run update-webdriver
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs juice-shop
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls juice-shop
npm ERR! There is likely additional logging output above.
[npm-debug.log](https://github.com/bkimminich/juice-shop/files/1475410/npm-debug.log)

npm ERR! Please include the following file with any support request:
npm ERR!     /juice-shop/npm-debug.log
/juice-shop #
```

The log file doesn't give any more info, see attached file. 
[npm-debug.log](https://github.com/bkimminich/juice-shop/files/1475417/npm-debug.log)

**Comment by J12934** [comment]
Are you using a system wide proxy?

I run the e2e test by running:
```bash
npm run preprotractor
./node_modules/.bin/protractor protractor.conf.js
```

And all the test were passing

**Comment by bkimminich** [comment]
@battletux  Are you trying `npm run protractor` from inside the container? That tries to start its own server in a spawned process for the test scripts and I'd guess it'd fail to do so. But if it works for @J12934 with your Docker image, I think we're good to go!

**Comment by battletux** [comment]
yes, I was being a bit dense and running it in the container.

I've set up protractor but it doesn't seem to want to work for me, it errors on `Can not find module 'config' `.

I dont want this to beat me, so I need to figure out what is wrong with the settings for protractor.

How ever if @J12934 says it works and your happy with that I'll send over the pull request.

**Comment by bkimminich** [comment]
@battletux You need to have the source code checked out and once grab all dependencies from NPM, so your commands should probably be

```
npm install
npm run preprotractor
./node_modules/.bin/protractor protractor.conf.js
```

This while having a Docker container running on port 3000 should do the trick.

**Comment by battletux** [comment]
@bkimminich That got it thanks! All passed :)

**Comment by bkimminich** [comment]
Merged and ready to be ditributed with v6.0.0. All `snapshot` containers will already benefit from this _now_!
